### ISSUE TITLE
System Environment Check

### ISSUE THREAD

**Issue by ntotten** [forcedotcom/salesforcedx-vscode#902]
Currently, when a user tries to run the Salesforce extensions and is missing one of the prerequisites on their computer we simply show an error message. This experience is causing confusion to developers and we should do a better job informing them of system setup problems.

The proposed solution is for the extensions to perform a "system" check in the background after the extension is activated. Any feature of an extension that requires a system dependency will be activated only after the system checks succeeds. 

In the event the system check fails we will show the user a custom webview providing them with specific setup instructions tailored to their operating system. The webview will have a button that allows the user to rerun the system check. If the test passes then the extension will fully activate, if it fails the webview will be updated to show the reason why it failed.

We can use this method to check for things like:
* If JRE is installed
* Is the version of JRE supported
* Is JRE correctly configured in the path
* Is the Salesforce CLI installed and correctly configured

This experience will dramatically help new users successfully install and use the extensions

WID-123455

**Comment by toanshulverma** [comment]
How about allowing these settings to be configured within settings file. Instead of current way, where it assumes that the installation path are added to environment settings?

For instance, if developer is not able to update PATH within environment settings, there is currently no way for SFDX extensions to use a custom path. If we can specify custom path in settings, then this could work.

**Comment by chexxor** [comment]
Would you accept a PR which adds a means of specifying the absolute path to the `sfdx` executable using a VS Code setting and/or an environment variable?

**Comment by chexxor** [comment]
Looks like we would need to change [sfdxCommandBuilder](https://github.com/forcedotcom/salesforcedx-vscode/blob/develop/packages/salesforcedx-utils-vscode/src/cli/commandBuilder.ts#L77-L81) to get the `'sfdx'` string from a configuration variable.

I imagine some people might want to set this value from an environment variable in the terminal which executes the sfdx commands, so we might want to abstract it out to a module of its own. I don't know if maintainers would prefer to avoid env vars, though, or would like to avoid a configuration class for this module. Looks like that "salesforcedx-utils-vscode" module doesn't have a configuration module yet, so we could make one.

**Comment by smaddox-sf** [comment]
Hi @chexxor - Your proposed PR doesn't seem to address creating a system environment check (which is the topic of this issue).  As for specifying the path, are you referring to the path to the CLI?  We have an existing environment variable for that - https://developer.salesforce.com/tools/vscode/en/troubleshooting/#set-salesforce-cli-path-windows

**Comment by chexxor** [comment]
Hi @smaddox-sf, thanks for the link!

This was the most relevant issue I could find about configuring the PATH environment variable. I was interested in configuring the PATH within this VS Code extension to ensure it uses a specific version of `sfdx`, leaving me free to use a different `sfdx` in different situations. This extension (or one in the greater umbrella project) already has a similar configuration variable named `salesforcedx-vscode-apex` for defining the location of the JVM, which also has a dedicated environment which you could use to accomplish the same thing.

Sorry to accidentally hijack this issue -- let me know if you're like me to create a different one for this!

**Comment by chexxor** [comment]
I created a new issue @smaddox-sf: https://github.com/forcedotcom/salesforcedx-vscode/issues/2446

I hope this doesn't side-track the original topic of the issue. Feel free to delete my comments to clear up this issue! Let me know if you need help!

**Comment by diyer** [comment]
Closing this one as we have some system requirements checks in place when installing extensions. Please file a new bug for new issues seen.
cc: @AnanyaJha
